src/mesh/mesh.hpp

```diff
@@ -73,25 +73,44 @@ KOKKOS_INLINE_FUNCTION void par_for_inner(const team_mbr_t &team_member, const i
 //----------------------------------------------------------------------------------------
 //! \class MeshBlock
 //  \brief data/functions associated with a single block
-class MeshBlock {
+class MeshBlock : public std::enable_shared_from_this<MeshBlock> {
   friend class RestartOutput;
   friend class Mesh;
 
  public:
+  MeshBlock() = default;
   MeshBlock(const int n_side, const int ndim); // for Kokkos testing with ghost
-  MeshBlock(int igid, int ilid, LogicalLocation iloc, RegionSize input_size,
-            BoundaryFlag *input_bcs, Mesh *pm, ParameterInput *pin,
-            ApplicationInput *app_in, Properties_t &properties, int igflag,
-            bool ref_flag = false);
-  MeshBlock(int igid, int ilid, LogicalLocation iloc, RegionSize input_block,
-            BoundaryFlag *input_bcs, Mesh *pm, ParameterInput *pin,
-            ApplicationInput *app_in, Properties_t &properties, Packages_t &packages,
-            int igflag, bool ref_flag = false);
-  MeshBlock(int igid, int ilid, Mesh *pm, ParameterInput *pin, ApplicationInput *app_in,
-            Properties_t &properties, Packages_t &packages, LogicalLocation iloc,
-            RegionSize input_block, BoundaryFlag *input_bcs, double icost, int igflag);
   ~MeshBlock();
 
+  // Initializer to set up a meshblock called with the default constructor
+  // This is necessary because the back pointers can't be set up until
+  // the block is allocated.
+  void Initialize(int igid, int ilid, LogicalLocation iloc, RegionSize input_block,
```

Comment:
Can this be private?

Reply:
If we always use the factory functions, yes. But I figured there might be cases where a non-factory call will be made.

Reply:
In general for API design, I prefer to limit new APIs by default, and only introduce them when need has arisen. The issue with adding APIs because you anticipate their usage is that they tend not to be built-to-fit.

Reply:
Yeah ok. I suspect something like this may need to be public eventually... or more factory functions may need to be created. But for now I'll make it private.